applications/server/modules/access.py

```diff
@@ -0,0 +1,83 @@
+from gluon import current
+from gluon import auth
+
+
+# Permission types.
+# v = view
+# a = admin (valid only for one whole device)
+# e = edit settings of procedure
+
+def register_device(name=None, description=''):
+    db = current.db
+    devices = db.devices
+    permission = db.user_permission
+    device_id = devices.insert(name=name, user_id=auth.user_id, description=description)
+    permission.insert(perm_user_email=auth.user.email, device_id=device_id, procedure_id=None, perm_type='a')
+
+
+# Share one device to multiple users with one permission type
+# return emails which are not successfully shared.
+def share_device(device_id, user_email=[], type='v', procedure_id=None):
+    db = current.db
+    permission = db.user_permission
+    fail_email = []
+    for email in user_email:
+        # Does the email exist in registered user table?
+        # if it does not exist this permission adding operation is denied
+        if db(db.auth_user.email == email) is None:
+            fail_email.append(email)
+        else:
+            # Does permission table contain this record?
+            p = db((db.user_permission.perm_user_email == email) &
+                   (db.user_permission.device_id == device_id) &
+                   (db.user_permission.procedure_id == procedure_id)).select().first()
+
+            if p is None:
+                # if no record stored in permission table we insert new permission record
+                permission.insert(perm_user_email=email, device_id=device_id, procedure_id=procedure_id, perm_type=type)
+            else:
+                # update the selected row
+                p.update(perm_user_email=email, device_id=device_id, procedure_id=procedure_id, perm_type=type)
```

Comment:
If you do it that way, if I have already shared a device with you in 'a' (admin) mode, and I now share it with you (perhaps as part of a group of more devices) in 'v' mode, you lose the 'a' access to this device. 
If you call a function share_device, then the function should only broaden, never restrict, access. 
So perhaps you should take the max of the new and old permissions.  So you should implement some functions that give permissions_max(x, y) and permissions_min(x, y), for permissions x, y.